man/mid3v2.1
.\" Man page generated from reStructuredText.
.
.TH MID3V2 1 "" "" ""
.SH NAME
mid3v2 \- audio tag editor similar to 'id3v2'
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.sp
\fBmid3v2\fP [\fIoptions\fP] \fIfilename\fP ...
.SH DESCRIPTION
.sp
\fBmid3v2\fP is a Mutagen\-based replacement for id3lib\(aqs id3v2. It supports
ID3v2.4 and more frames; it also does not have the numerous bugs that plague
id3v2.
.sp
This program exists mostly for compatibility with programs that want to tag
files using id3v2. For a more usable interface, we recommend Ex Falso.
.SH OPTIONS
.INDENT 0.0
.TP
.B \-q\fP,\fB  \-\-quiet
Be quiet: do not mention file operations that perform the user\(aqs
request. Warnings will still be printed.
.TP
.B \-v\fP,\fB  \-\-verbose
Be verbose: state all operations performed. This is the opposite of
\-\-quiet. This is the default.
.TP
.B \-e\fP,\fB  \-\-escape
Enable interpretation of backslash escapes for tag values.
Makes it possible to escape the colon\-separator in TXXX, COMM
values like \(aq\e:\(aq and insert escape sequences like \(aq\en\(aq, \(aq\et\(aq etc.
.TP
.B \-f\fP,\fB  \-\-list\-frames
Display all supported ID3v2.3/2.4 frames and their meanings.
.TP
.B \-L\fP,\fB  \-\-list\-genres
List all ID3v1 numeric genres. These can be used to set TCON frames,
but it is not recommended.
.TP
.B \-l\fP,\fB  \-\-list
List all tags in the files. The output format is \fInot\fP the same as
id3v2\(aqs; instead, it is easily parsable and readable. Some tags may not
have human\-readable representations.
.TP
.B \-\-list\-raw
List all tags in the files, in raw format. Although this format is
nominally human\-readable, it may be very long if the tag contains
embedded binary data.
.TP
.B \-d\fP,\fB  \-\-delete\-v2
Delete ID3v2 tags.
.TP
.B \-s\fP,\fB  \-\-delete\-v1
Delete ID3v1 tags.
.TP
.B \-D\fP,\fB  \-\-delete\-all
Delete all ID3 tags.
.TP
.BI \-\-delete\-frames\fB= FRAMES
Delete specific ID3v2 frames (or groups of frames) from the files.
\fBFRAMES\fP is a "," separated list of frame names e.g. \fB"TPE1,TALB"\fP
.TP
.B \-C\fP,\fB  \-\-convert
Convert ID3v1 tags to ID3v2 tags. This  will also happen automatically
during any editing.
.TP
.BI \-a\fP,\fB  \-\-artist\fB= ARTIST
Set the artist information (TPE1).
.TP
.BI \-A\fP,\fB  \-\-album\fB= ALBUM
Set the album information (TALB).
.TP
.BI \-t\fP,\fB  \-\-song\fB= TITLE
Set the title information (TIT2).
.TP
.BI \-c\fP,\fB  \-\-comment\fB= <DESCRIPTION:COMMENT:LANGUAGE>
Set a comment (COMM). The language and description may be omitted, in
which case the language defaults to English, and the description to an
empty string.
.TP
.BI \-p\fP,\fB  \-\-picture\fB= <FILENAME:DESCRIPTION:IMAGE\-TYPE:MIME\-TYPE>
Set the attached picture (APIC). Everything except the filename can be
omitted in which case default values will be used.
.TP
.BI \-g\fP,\fB  \-\-genre\fB= GENRE
Set the genre information (TCON).
.TP
.BI \-y\fP,\fB  \-\-year\fB= <YYYY>\fP,\fB \ \-\-date\fB= <YYYY\-[MM\-DD]>
Set the year/date information (TDRC).
.TP
.BI \-T\fP,\fB  \-\-track\fB= <NUM/NUM>
Set the track number (TRCK).
.UNINDENT
.sp
Any text or URL frame (those beginning with T or W) can be modified or
added by prefixing the name of the frame with "\-\-". For example, \fB\-\-TIT3
"Monkey!"\fP will set the TIT3 (subtitle) frame to \fBMonkey!\fP\&.
.sp
The TXXX frame requires a colon\-separated description key; many TXXX frames
may be set in the file as long as they have different keys. To set this
key, just separate the text with a colon, e.g. \fB\-\-TXXX
"ALBUMARTISTSORT:Examples, The"\fP\&.
.sp
The USLT frame has the format <DESCRIPTION:TEXT:LANGUAGE>. The language and
description may be omitted, in which case the language defaults to English,
and the description to an empty string.
.sp
The special POPM frame can be set in a similar way: \fB\-\-POPM
"bob@example.com:128:2"\fP to set Bob\(aqs rating to 128/255 with 2 plays.
.SH BUGS
.sp
No sanity checking is done on the editing operations you perform, so mid3v2
will happily accept \-\-TSIZ when editing an ID3v2.4 frame. However, it will
also automatically throw it out during the next edit operation.
.SH AUTHOR
.sp
Joe Wreschnig is the author of mid3v2, but he doesn\(aqt like to admit it.
.\" Generated by docutils manpage writer.
.


NEWS
1.37 - 2017.02.24
-----------------

* Relicense "GPLv2" → "GPLv2 or later" :bug:`291`
* DSF: add `mutagen.dsf` module for DSF (DSD Stream File) support
  :pr:`283` (Boris Pruessmann)
* MP3: Add `mp3.MPEGInfo.encoder_settings` containing a guess of the encoder
  settings used, for example ``"-V2"`` for LAME :bug:`66`
* ID3: add iTunes movement related frames `id3.MVIN` and `id3.MVNM`
* MP4: support ``©mvi``, ``©mvc``, ``shwm``, ``stik``, ``rtng``, ``tves``,
  ``tvsn``, ``plID``, ``cnID``, ``geID``, ``atID``, ``sfID``, ``cmID``,
  ``akID`` :bug:`130`


1.36.3 - 2017.02.24
-------------------

* MP3: fix error with xing frames without a frame count :bug:`292`


1.36.2 - 2017.01.25
-------------------

* ID3: Always write little endian utf-16 with BOM.
  Fixes tests on big endian machines :pr:`289`


1.36.1 - 2017.01.22
-------------------

* Support GAE runtime :bug:`286`
* FLAC: Fix crash when loading files with zero samples :bug:`287`
* MP3: Handle broken lame tags written by older lame versions


1.36 - 2016.12.22
-----------------

* ID3: Ignore trailing empty values for v2.3 text frames :bug:`276`
* ID3: Write large APIC frames last :bug:`278`
* EasyID3: support saving as v2.3 :bug:`188`
* FLAC: Add StreamInfo.bitrate :bug:`279`
* mid3cp: Add ``--merge`` option :bug:`277`
* MP4: Allow loading files without audio tracks :bug:`272`


1.35.1 - 2016.11.09
-------------------

* Revert back to distutils :bug:`273`


1.35 - 2016.11.02
-----------------

* Tests: Require pytest
* Tools: Install .exe launchers on Windows
* setup.py: Require setuptools
* ID3:

  * Fix loading files with CRM frames :bug:`239`
  * Fix loading AENC, LINK, GRID frames with no payload
  * Merge duplicate text frames with same key on load :bug:`172`
  * Allow parsing of duplicate APIC frames :bug:`172`
  * Parse utf-16 text fields with missing BOM :bug:`267`
  * Increase max resyncs for the mpeg frame search :bug:`268`


1.34.1 - 2016.08.13
-------------------

* ID3: Expose some internals again to make Picard (mostly) work again.

  * http://tickets.musicbrainz.org/browse/PICARD-833
  * https://github.com/metabrainz/picard/pull/479


1.34 - 2016.07.20
-----------------

* ID3:

  * Add `CTOC <id3.CTOC>` and `CHAP <id3.CHAP>` frames. New classes:
    `ID3Tags <id3.ID3Tags>`, `CTOCFlags <id3.CTOCFlags>`. :bug:`6`
  * Add `TCAT <id3.TCAT>`, `TKWD <id3.TKWD>`, `PCST <id3.PCST>` frames.
    :bug:`249`
  * Validate user provided LNK/LINK frameid. :bug:`242`
  * Add `RVAD <id3.RVAD>`, RVA frames
  * Add TST, TSA, TS2, TSP and TSC frames
  * Fix not writing optional fields when saving to v2.3
  * Add default field values for all frames

* Drop Python 2.6 support
* EasyID3: Fix TXXX frame encoding when setting a non-latin1 encodable
  value after a latin1 one. :bug:`263`


1.33.2 - 2016.07.05
-------------------

* Fix loading of small ogg/apev2 files (1.33 regression)


1.33.1 - 2016.06.29
-------------------

* Fix Overeager deprecation warnings :bug:`261`


1.33 - 2016.06.29
-----------------

* FileType, Metadata: File-like object support :bug:`1`
* mid3v2: Add APIC support. :bug:`47`
* EasyID3: Fix handling of RVA2 frames with non-lowercase description
  :bug:`215`
* mid3v2: Add UFID support. :bug:`234`
* ID3: Include human-readable representation of the picture type in
  APIC._pprint() :bug:`244`
* EasyID3: make albumartist use TPE2 and move performer to TXXX. :bug:`252`
* ID3: id3.ID3TimeStamp comparator: check type :pr:`260`
  (Fabian Peter Hammerle)
* setup.py: follow PEP440 for the development version
* FileType/Metadata.load/save/delete no longer raise IOError or IOError
  subclasses. They only raise subclasses of MutagenError.


1.32 - 2016.05.02
-----------------

* Add basic SMF (Standard MIDI File) support (:mod:`mutagen.smf`)
* FLAC: add ``audio/flac`` mime type. :bug:`235`
* ASF: Fixed crash when object size is longer than the header and file length
  (Ben Ockmore)
* ID3: Validate attributes set after frame creation :commit:`69368c31e00`
  (:user:`Daniel Plachotich <danpla>`)
* MP4: validate values in ``__setitem__`` so things don't fail in save()
  :bug:`236`
* tests: Fix SynchronizedTextSpec test on big-endian machines :bug:`247`
  (Daniel Plachotich)
* ID3: do type checking in ``__setitem__`` :bug:`251`
* Building the documentation now requires sphinx >= 1.3
* New :class:`mutagen.Tags` base class for tags
* Moved from Bitbucket to GitHub


1.31 - 2015.09.10
-----------------

* New padding control API for flac/id3/mp4/asf/ogg/aiff and everything
  based on it, except oggflac :bug:`229`
* Mutagen will now reduce padding on save if there is lots of it.
* delete() will remove padding in addition to tags.

* ASF:

  * Padding support :bug:`201`
  * Don't report negative lengths for some broken files
  * New :class:`asf.ASFInfo`.codec_type/codec_name/codec_description
  * Implement ASF.delete()

* OGG: Padding support for Opus/Vorbis/Theora/Speex
* M4A: Implementation removed. Every operation will raise. Use mp4 instead.
* Tools: Support Unicode output under Windows :bug:`232`


1.30 - 2015.08.22
-----------------

* FLAC:

  * Fix :meth:`flac.FLAC.save` in case the source contained a too large
    (invalid but recovered) image block :bug:`226`

* MP3:

  * Improved length and bitrate accuracy:

    * Read lame "Info" tags for improved bitrate/length accuracy
    * Use bytes info of VBRI headers for improved bitrate accuracy
    * Subtract encoder delay/padding from length for improved length accuracy
      (especially for short tracks)
    * Fix rare false identification of Xing headers :bug:`182`

  * New :class:`mp3.MPEGInfo`.encoder_info attribute containing the encoder
    name and version :bug:`66`
  * New :class:`mp3.MPEGInfo`.bitrate_mode attribute exposing if the file is
    VBR, ABR or CBR :bug:`24` :bug:`66`
  * New :class:`mp3.MPEGInfo`.channels attribute providing the channel count
  * New :class:`mp3.MPEGInfo`.track_gain/track_peak/album_gain values exposing

mutagen/_tools/mid3v2.py
# -*- coding: utf-8 -*-
# Copyright 2005 Joe Wreschnig
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

"""Pretend to be /usr/bin/id3v2 from id3lib, sort of."""

import sys
import codecs
import mimetypes

from optparse import SUPPRESS_HELP

import mutagen
import mutagen.id3
from mutagen.id3 import Encoding, PictureType
from mutagen._senf import fsnative, print_, argv, fsn2text, fsn2bytes, \
    bytes2fsn
from mutagen._compat import PY2, text_type

from ._util import split_escape, SignalHandler, OptionParser


VERSION = (1, 3)
_sig = SignalHandler()

global verbose
verbose = True


class ID3OptionParser(OptionParser):
    def __init__(self):
        mutagen_version = ".".join(map(str, mutagen.version))
        my_version = ".".join(map(str, VERSION))
        version = "mid3v2 %s\nUses Mutagen %s" % (my_version, mutagen_version)
        self.edits = []
        OptionParser.__init__(
            self, version=version,
            usage="%prog [OPTION] [FILE]...",
            description="Mutagen-based replacement for id3lib's id3v2.")

    def format_help(self, *args, **kwargs):
        text = OptionParser.format_help(self, *args, **kwargs)
        return text + """\
You can set the value for any ID3v2 frame by using '--' and then a frame ID.
For example:
        mid3v2 --TIT3 "Monkey!" file.mp3
would set the "Subtitle/Description" frame to "Monkey!".

Any editing operation will cause the ID3 tag to be upgraded to ID3v2.4.
"""


def list_frames(option, opt, value, parser):
    items = mutagen.id3.Frames.items()
    for name, frame in sorted(items):
        print_(u"    --%s    %s" % (name, frame.__doc__.split("\n")[0]))
    raise SystemExit


def list_frames_2_2(option, opt, value, parser):
    items = mutagen.id3.Frames_2_2.items()
    items.sort()
    for name, frame in items:
        print_(u"    --%s    %s" % (name, frame.__doc__.split("\n")[0]))
    raise SystemExit


def list_genres(option, opt, value, parser):
    for i, genre in enumerate(mutagen.id3.TCON.GENRES):
        print_(u"%3d: %s" % (i, genre))
    raise SystemExit


def delete_tags(filenames, v1, v2):
    for filename in filenames:
        with _sig.block():
            if verbose:
                print_(u"deleting ID3 tag info in", filename, file=sys.stderr)
            mutagen.id3.delete(filename, v1, v2)


def delete_frames(deletes, filenames):

    try:
        deletes = frame_from_fsnative(deletes)
    except ValueError as err:
        print_(text_type(err), file=sys.stderr)

    frames = deletes.split(",")

    for filename in filenames:
        with _sig.block():
            if verbose:
                print_(u"deleting %s from" % deletes, filename,
                       file=sys.stderr)
            try:
                id3 = mutagen.id3.ID3(filename)
            except mutagen.id3.ID3NoHeaderError:
                if verbose:
                    print_(u"No ID3 header found; skipping.", file=sys.stderr)
            except Exception as err:
                print_(text_type(err), file=sys.stderr)
                raise SystemExit(1)
            else:
                for frame in frames:
                    id3.delall(frame)
                id3.save()


def frame_from_fsnative(arg):
    """Takes item from argv and returns ascii native str
    or raises ValueError.
    """

    assert isinstance(arg, fsnative)

    text = fsn2text(arg, strict=True)
    if PY2:
        return text.encode("ascii")
    else:
        return text.encode("ascii").decode("ascii")


def value_from_fsnative(arg, escape):
    """Takes an item from argv and returns a text_type value without
    surrogate escapes or raises ValueError.
    """

    assert isinstance(arg, fsnative)

    if escape:
        bytes_ = fsn2bytes(arg, "utf-8")
        if PY2:
            bytes_ = bytes_.decode("string_escape")
        else:
            bytes_ = codecs.escape_decode(bytes_)[0]
        arg = bytes2fsn(bytes_, "utf-8")

    text = fsn2text(arg, strict=True)
    return text


def error(*args):
    print_(*args, file=sys.stderr)
    raise SystemExit(1)


def get_frame_encoding(frame_id, value):
    if frame_id == "APIC":
        # See https://github.com/beetbox/beets/issues/899#issuecomment-62437773
        return Encoding.UTF16 if value else Encoding.LATIN1
    else:
        return Encoding.UTF8


def write_files(edits, filenames, escape):
    # unescape escape sequences and decode values
    encoded_edits = []
    for frame, value in edits:
        if not value:
            continue

        try:
            frame = frame_from_fsnative(frame)
        except ValueError as err:
            print_(text_type(err), file=sys.stderr)

        assert isinstance(frame, str)

        # strip "--"
        frame = frame[2:]

        try:
            value = value_from_fsnative(value, escape)
        except ValueError as err:
            error(u"%s: %s" % (frame, text_type(err)))

        assert isinstance(value, text_type)

        encoded_edits.append((frame, value))
    edits = encoded_edits

    # preprocess:
    #   for all [frame,value] pairs in the edits list
    #      gather values for identical frames into a list
    tmp = {}
    for frame, value in edits:
        if frame in tmp:
            tmp[frame].append(value)
        else:
            tmp[frame] = [value]
    # edits is now a dictionary of frame -> [list of values]
    edits = tmp

    # escape also enables escaping of the split separator
    if escape:
        string_split = split_escape
    else:
        string_split = lambda s, *args, **kwargs: s.split(*args, **kwargs)

    for filename in filenames:
        with _sig.block():
            if verbose:
                print_(u"Writing", filename, file=sys.stderr)
            try:
                id3 = mutagen.id3.ID3(filename)
            except mutagen.id3.ID3NoHeaderError:
                if verbose:
                    print_(u"No ID3 header found; creating a new tag",
                          file=sys.stderr)
                id3 = mutagen.id3.ID3()
            except Exception as err:
                print_(str(err), file=sys.stderr)
                continue
            for (frame, vlist) in edits.items():
                if frame == "POPM":
                    for value in vlist:
                        values = string_split(value, ":")
                        if len(values) == 1:
                            email, rating, count = values[0], 0, 0
                        elif len(values) == 2:
                            email, rating, count = values[0], values[1], 0
                        else:
                            email, rating, count = values

                        frame = mutagen.id3.POPM(
                            email=email, rating=int(rating), count=int(count))
                        id3.add(frame)
                elif frame == "APIC":
                    for value in vlist:
                        values = string_split(value, ":")
                        # FIXME: doesn't support filenames with an invalid
                        # encoding since we have already decoded at that point
                        fn = values[0]

                        if len(values) >= 2:
                            desc = values[1]
                        else:
                            desc = u"cover"

                        if len(values) >= 3:
                            try:
                                picture_type = int(values[2])
                            except ValueError:
                                error(u"Invalid picture type: %r" % values[1])
                        else:
                            picture_type = PictureType.COVER_FRONT

                        if len(values) >= 4:
                            mime = values[3]
                        else:
                            mime = mimetypes.guess_type(fn)[0] or "image/jpeg"

                        if len(values) >= 5:
                            error("APIC: Invalid format")

                        encoding = get_frame_encoding(frame, desc)

                        try:
                            with open(fn, "rb") as h:
                                data = h.read()
                        except IOError as e:
                            error(text_type(e))

                        frame = mutagen.id3.APIC(encoding=encoding, mime=mime,
                            desc=desc, type=picture_type, data=data)

                        id3.add(frame)
                elif frame == "COMM":
                    for value in vlist:
                        values = string_split(value, ":")
                        if len(values) == 1:
                            value, desc, lang = values[0], "", "eng"
                        elif len(values) == 2:
                            desc, value, lang = values[0], values[1], "eng"
                        else:
                            value = ":".join(values[1:-1])
                            desc, lang = values[0], values[-1]
                        frame = mutagen.id3.COMM(
                            encoding=3, text=value, lang=lang, desc=desc)
                        id3.add(frame)
                elif frame == "USLT":
                    for value in vlist:
                        values = string_split(value, ":")
                        if len(values) == 1:
                            value, desc, lang = values[0], "", "eng"
                        elif len(values) == 2:
                            desc, value, lang = values[0], values[1], "eng"
                        else:
                            value = ":".join(values[1:-1])
                            desc, lang = values[0], values[-1]
                        frame = mutagen.id3.USLT(
                            encoding=3, text=value, lang=lang, desc=desc)
                        id3.add(frame)
                elif frame == "UFID":
                    for value in vlist:
                        values = string_split(value, ":")
                        if len(values) != 2:
                            error(u"Invalid value: %r" % values)
                        owner = values[0]
                        data = values[1].encode("utf-8")
                        frame = mutagen.id3.UFID(owner=owner, data=data)
                        id3.add(frame)
                elif frame == "TXXX":
                    for value in vlist:
                        values = string_split(value, ":", 1)
                        if len(values) == 1:
                            desc, value = "", values[0]
                        else:
                            desc, value = values[0], values[1]
                        frame = mutagen.id3.TXXX(
                            encoding=3, text=value, desc=desc)
                        id3.add(frame)
                elif issubclass(mutagen.id3.Frames[frame],
                                mutagen.id3.UrlFrame):
                    frame = mutagen.id3.Frames[frame](encoding=3, url=vlist)
                    id3.add(frame)
                else:
                    frame = mutagen.id3.Frames[frame](encoding=3, text=vlist)
                    id3.add(frame)
            id3.save(filename)


def list_tags(filenames):
    for filename in filenames:
        print_("IDv2 tag info for", filename)
        try:
            id3 = mutagen.id3.ID3(filename, translate=False)
        except mutagen.id3.ID3NoHeaderError:
            print_(u"No ID3 header found; skipping.")
        except Exception as err:
            print_(text_type(err), file=sys.stderr)
            raise SystemExit(1)
        else:
            print_(id3.pprint())


def list_tags_raw(filenames):
    for filename in filenames:
        print_("Raw IDv2 tag info for", filename)
        try:
            id3 = mutagen.id3.ID3(filename, translate=False)
        except mutagen.id3.ID3NoHeaderError:
            print_(u"No ID3 header found; skipping.")
        except Exception as err:
            print_(text_type(err), file=sys.stderr)
            raise SystemExit(1)
        else:
            for frame in id3.values():
                print_(text_type(repr(frame)))


def main(argv):
    parser = ID3OptionParser()
    parser.add_option(
        "-v", "--verbose", action="store_true", dest="verbose", default=False,
        help="be verbose")
    parser.add_option(
        "-q", "--quiet", action="store_false", dest="verbose",
        help="be quiet (the default)")
    parser.add_option(
        "-e", "--escape", action="store_true", default=False,
        help="enable interpretation of backslash escapes")
    parser.add_option(
        "-f", "--list-frames", action="callback", callback=list_frames,
        help="Display all possible frames for ID3v2.3 / ID3v2.4")
    parser.add_option(
        "--list-frames-v2.2", action="callback", callback=list_frames_2_2,
        help="Display all possible frames for ID3v2.2")
    parser.add_option(
        "-L", "--list-genres", action="callback", callback=list_genres,
        help="Lists all ID3v1 genres")
    parser.add_option(
        "-l", "--list", action="store_const", dest="action", const="list",
        help="Lists the tag(s) on the open(s)")
    parser.add_option(
        "--list-raw", action="store_const", dest="action", const="list-raw",
        help="Lists the tag(s) on the open(s) in Python format")
    parser.add_option(
        "-d", "--delete-v2", action="store_const", dest="action",
        const="delete-v2", help="Deletes ID3v2 tags")
    parser.add_option(
        "-s", "--delete-v1", action="store_const", dest="action",
        const="delete-v1", help="Deletes ID3v1 tags")
    parser.add_option(
        "-D", "--delete-all", action="store_const", dest="action",
        const="delete-v1-v2", help="Deletes ID3v1 and ID3v2 tags")
    parser.add_option(
        '--delete-frames', metavar='FID1,FID2,...', action='store',
        dest='deletes', default='', help="Delete the given frames")
    parser.add_option(
        "-C", "--convert", action="store_const", dest="action",
        const="convert",
        help="Convert tags to ID3v2.4 (any editing will do this)")

    parser.add_option(
        "-a", "--artist", metavar='"ARTIST"', action="callback",
        help="Set the artist information", type="string",
        callback=lambda *args: args[3].edits.append((fsnative(u"--TPE1"),
                                                     args[2])))
    parser.add_option(
        "-A", "--album", metavar='"ALBUM"', action="callback",
        help="Set the album title information", type="string",
        callback=lambda *args: args[3].edits.append((fsnative(u"--TALB"),
                                                     args[2])))
    parser.add_option(
        "-t", "--song", metavar='"SONG"', action="callback",
        help="Set the song title information", type="string",
        callback=lambda *args: args[3].edits.append((fsnative(u"--TIT2"),
                                                     args[2])))
    parser.add_option(
        "-c", "--comment", metavar='"DESCRIPTION":"COMMENT":"LANGUAGE"',
        action="callback", help="Set the comment information", type="string",
        callback=lambda *args: args[3].edits.append((fsnative(u"--COMM"),
                                                     args[2])))
    parser.add_option(
        "-p", "--picture",
        metavar='"FILENAME":"DESCRIPTION":"IMAGE-TYPE":"MIME-TYPE"',
        action="callback", help="Set the picture", type="string",
        callback=lambda *args: args[3].edits.append((fsnative(u"--APIC"),
                                                     args[2])))
    parser.add_option(
        "-g", "--genre", metavar='"GENRE"', action="callback",
        help="Set the genre or genre number", type="string",
        callback=lambda *args: args[3].edits.append((fsnative(u"--TCON"),
                                                     args[2])))
    parser.add_option(
        "-y", "--year", "--date", metavar='YYYY[-MM-DD]', action="callback",
        help="Set the year/date", type="string",
        callback=lambda *args: args[3].edits.append((fsnative(u"--TDRC"),
                                                     args[2])))
    parser.add_option(
        "-T", "--track", metavar='"num/num"', action="callback",
        help="Set the track number/(optional) total tracks", type="string",
        callback=lambda *args: args[3].edits.append((fsnative(u"--TRCK"),
                                                     args[2])))

    for key, frame in mutagen.id3.Frames.items():
        if (issubclass(frame, mutagen.id3.TextFrame)
                or issubclass(frame, mutagen.id3.UrlFrame)
                or issubclass(frame, mutagen.id3.POPM)
                or frame in (mutagen.id3.APIC, mutagen.id3.UFID,
                             mutagen.id3.USLT)):
            parser.add_option(
                "--" + key, action="callback", help=SUPPRESS_HELP,
                type='string', metavar="value",  # optparse blows up with this
                callback=lambda *args: args[3].edits.append(args[1:3]))

    (options, args) = parser.parse_args(argv[1:])
    global verbose
    verbose = options.verbose

    if args:
        if parser.edits or options.deletes:
            if options.deletes:
                delete_frames(options.deletes, args)
            if parser.edits:
                write_files(parser.edits, args, options.escape)
        elif options.action in [None, 'list']:
            list_tags(args)
        elif options.action == "list-raw":
            list_tags_raw(args)
        elif options.action == "convert":
            write_files([], args, options.escape)
        elif options.action.startswith("delete"):
            delete_tags(args, "v1" in options.action, "v2" in options.action)
        else:
            parser.print_help()
    else:
        parser.print_help()


def entry_point():
    _sig.init()
    return main(argv)
